lib/galaxy/jobs/runners/local.py

```diff
@@ -214,10 +214,16 @@ def _check_pid(self, pid):
             return False
 
     def _terminate(self, proc):
-        os.killpg(proc.pid, 15)
-        sleep(1)
-        if proc.poll() is None:
-            os.killpg(proc.pid, 9)
+        process_exists = True
+        try:
+            os.killpg(proc.pid, 15)
+        except OSError as e:
+            if e.errno == errno.ESRCH:
+                process_exists = False
+        if process_exists:
+            sleep(1)
+            if proc.poll() is None:
```

Comment:
Given the discussion above, this is not consistent with working on a process group, given that `proc.poll()` only checks if the "main" child process has terminated.

Reply:
> This cold use `os.killpg(proc.pid, 0)` I believe?

How? It doesn't seem to have any effect. I would just do something like:
```python
            try:
                os.killpg(proc.pid, signal.SIGTERM)
            except OSError as e:
                if e.errno == errno.ESRCH:
                    pass
```

Reply:
> Will it raise `ESRCH` if the original process is dead but other processes in the group are still running?

No, it doesn't.